src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Key.java

```diff
@@ -1533,6 +1533,8 @@ void releaseKeyID() {
                     // destroy
                     this.keyID = 0;
                     this.ref.removeNativeKey();
+                    // prevent enqueuing SessionKeyRef until removeNativeKey is done
+                    Reference.reachabilityFence(this);
```

Comment:
The approach we are now taking is to put the reachabilityFence() call within the finally-clause of a try-finally statement. This ensures that all paths through the method will pass through the reachability fence, regardless of inlining or other JIT optimizations.